### PR TITLE
Switch to kernels from github.com/lf-edge/eve-kernel

### PR DESCRIPTION
**Pull request by mikem-zed**
In this PR only x86-generic flavor is covered. We need to fix GH action in eve-kernel repo to actually push images to dockerhub (currently we miss a required secret).  After this issue is fixed I'll replicate GH actions to all other branches and update commit hashes in Makefile.
@yash-zededa 's private docker hub repo is used in this PR. It also has to be fixed

@rene could you please check the logic related to handling PLATFORM variable? I'm not sure whether I have to add handling of imx8 targets now or later

**UPDATE:** now all platforms are covered. Just need to change commit IDs to real ones

**UPDATE 2**: all commit numbers are now set to real values

### CODE REVIEW

**Comment by rouming** [comment]
@mikem-zed could you please add more explicit description to the PR: what works, what does not. Arm? Preempt-rt? Were all patches merged to the eve-kernel branches? Just to understand what other work is required in the future. Or it might be everything is done, then "everything was fully ported and compiled on all platforms" sentence works as well. Thanks.

File: Makefile
Comment (eriknordmark) [inline Makefile:147]: The KERNEL_BRANCH starts with eve-kernel, which means that the tag would do the same. But the tag is presumably appended as lfedge/eve-kernel:<tag>, meaning that eve-kernel would be repeated. That doesn't make sense.

File: Makefile
Comment (eriknordmark) [inline Makefile:151]: I think docker calls this a name; definitely not a tag.

File: Makefile
Comment (eriknordmark) [inline Makefile:708]: Does this mean that eve-kernel will build and publish the kernel sbom?

**Review by eriknordmark — COMMENTED** [review]
Some comments/questions

File: Makefile
Comment (rene) [reply Makefile:151]: I agree with @eriknordmark , but if you look on all variables defined by the `tools/parse-pkgs.sh` they will also contain the name of the container + the tag, although they are called ???_TAG.... I think that's because of linuxkit, since `linuxkit pkg show-tag` shows the name of the container + its tag.... so IMO it's fine to let KERNEL_TAG....

Comment (deitch) [reply Makefile:151]: These are tags. Anything after the `:` in a container image name is a tag. What the contents of that tag are (as long as legitimate) is fair play and is part of the tag. `lkt pkg show-tag` uses the git tree hash, but to make a tag.

File: Makefile
Comment (rene) [inline Makefile:31]: @mikem-zed , you need to handle PLATFORM for ARM devices.... so far we might have the following platforms:

imx8mp_pollux
imx8mp_epc_r3720
imx8mq_evk

These should point to the https://github.com/lf-edge/eve-kernel/tree/eve-kernel-arm64-v6.1.38-generic branch

Soon we will have the nvidia, or probably nvidia-lenovo-se70, that should point to nvidia's kernel, so I don't think we should handle these platforms in the Makefile... we have a script under tools/prepare-platform.sh that process the platform and copy some files when needed... maybe you can add a parameter to return the corresponding kernel branch according to the platform, so in the Makefile you just call the script.... in this way every time we add a new platform, we change only the "platform" script and not the Makefile...

Comment (rene) [reply Makefile:31]: or you could make a new script to handle all kernel branches according to the platform, I think it would make more sense...

File: Makefile
Comment (mikem-zed) [reply Makefile:147]: the resulting image name looks like this `docker.io/lfedge/eve-kernel/eve-kernel-amd64-v6.1.38-generic-29700903aee7-gcc` tag repeats branch name because all docker images are stored in the same repo docker.io/lfedge/eve-kernel. For easier troubleshooting it is nice to see full branch name in `images/rootfs.yml`

File: Makefile
Comment (mikem-zed) [reply Makefile:708]: yes, I have a jira ticket for it

File: Makefile
Comment (eriknordmark) [reply Makefile:147]: OK

File: Makefile
Comment (eriknordmark) [reply Makefile:151]: @deitch My comment was about KERNEL_TAG - not about KERNEL_DOCKER_TAG. Latter is a tag after the colon.
But since parse-pkgs has been calling the whole name+tag a tag forever, this is not the time and place to change that.

File: Makefile
Comment (eriknordmark) [reply Makefile:708]: Presumably we need that done before this goes in master so we can continue cutting EVE releases with a complete SBOM any time we want.

File: Makefile
Comment (deitch) [reply Makefile:151]: Ah, I had misunderstood. Thank you for correcting.

Yeah, the whole "is tag the whole thing or just the stuff after the `:`" is a painful nomenclature problem that has been around forever, and not just in eve. But this is not the place to fix it, agreed.

File: Makefile
Comment (mikem-zed) [reply Makefile:708]: I have created a PR https://github.com/lf-edge/eve/pull/3487 to build DOCKERFILE_ADD_SCANNER tool as a container. We need to push it to docker hub and then use in Makefile.eve to generate spdx.json file. When the kernel will be unpacked this spdx.json file will be picked up automatically by SYFT

**Review by eriknordmark — COMMENTED** [review]
Yash and I got the eve-kernel publish.yml to push. It pushed [kernel-amd64-v6.1.38-generic-9f4ce3ddb-gcc](https://hub.docker.com/layers/lfedge/eve-kernel/eve-kernel-amd64-v6.1.38-generic-9f4ce3ddb-gcc/images/sha256-2c7df7168541774492cd8cef951c2e817089dc4d680c1db2331c2c661cf43a55?context=explore)

Is that the tag you expect? When I ran it manually I got a longer hash in the tag.

**Comment by mikem-zed** [comment]
> Yash and I got the eve-kernel publish.yml to push. It pushed [kernel-amd64-v6.1.38-generic-9f4ce3ddb-gcc](https://hub.docker.com/layers/lfedge/eve-kernel/eve-kernel-amd64-v6.1.38-generic-9f4ce3ddb-gcc/images/sha256-2c7df7168541774492cd8cef951c2e817089dc4d680c1db2331c2c661cf43a55?context=explore)
> 
> Is that the tag you expect? When I ran it manually I got a longer hash in the tag.

yes, this tag looks as expected. I expect to see a short commit ID as hash here

File: Makefile
Comment (eriknordmark) [reply Makefile:31]: @rene what needs to be changed in this PR to make it usable (now or later) for ARM builds?

**Review by eriknordmark — APPROVED** [review]
LGTM but want confirmation from @rene and the SBOM generation in eve-kernel before we merge this.
Kicking off tests

**Comment by deitch** [comment]
SBoM generation for the kernel will require some restructuring. But that's small and self contained compared to redoing the kernel sourcing and testing itself. Don't hold up for that. I'll get to it next week.

**Comment by mikem-zed** [comment]
> SBoM generation for the kernel will require some restructuring. But that's small and self contained compared to redoing the kernel sourcing and testing itself. Don't hold up for that. I'll get to it next week.

Ok, then I'll fix @rene 's comments and resubmit the PR. But as I understand we need SBOM to do official releases of EVE. @eriknordmark what's your opinion on that?

**Comment by eriknordmark** [comment]
> SBoM generation for the kernel will require some restructuring. But that's small and self contained compared to redoing the kernel sourcing and testing itself. Don't hold up for that. I'll get to it next week.

@deitch We're scheduled to cut the next EVE build on Monday so it makes sense to hold this off until after that to avoid not having a complete SBoM for that build. OK?

**Comment by deitch** [comment]
Of course. I'll start work and get PR in when ready. We can choose to merge or not per the schedule you think is best.

**Comment by deitch** [comment]
Hey @mikem-zed help me out here.

In the current (pre-this-repo) eve kernel [Dockerfile](https://github.com/lf-edge/eve/blob/2192b043e47edbf095f0aac0d0da3531385001b5/pkg/kernel/Dockerfile), we have a few places where we add sources via `ADD`:

* [kernel source](https://github.com/lf-edge/eve/blob/2192b043e47edbf095f0aac0d0da3531385001b5/pkg/kernel/Dockerfile#L103) - downloaded from https://www.kernel.org/pub/linux/kernel/....
* [zfs driver](https://github.com/lf-edge/eve/blob/2192b043e47edbf095f0aac0d0da3531385001b5/pkg/kernel/Dockerfile#L240) - downloaded from https://github.com/openzfs/zfs
* [brektrou driver](https://github.com/lf-edge/eve/blob/2192b043e47edbf095f0aac0d0da3531385001b5/pkg/kernel/Dockerfile#L263C1-L263C5) - downloaded from https://github.com/brektrou/rtl8821CU

In the new [eve-kernel Dockerfile](https://github.com/lf-edge/eve-kernel/blob/eve-kernel-amd64-v6.1.38-generic/Dockerfile.gcc), I see:

* [zfs](https://github.com/lf-edge/eve-kernel/blob/eve-kernel-amd64-v6.1.38-generic/Dockerfile.gcc#L105)

That is it.

1. I assume the kernel source _is_ this repository?
2. Where is the brektrou driver? I don't see it in the repo at all? Do we no longer need it? Or is it sourced elsewhere?

**Comment by deitch** [comment]
@eriknordmark I would like your input on how we should add the kernel itself into the SBoM. Currently, because we download it as a blob, we have:

```json
    {
      "name": "linux-5.10.186.tar.xz",
      "SPDXID": "SPDXRef-linux-5.10.186.tar.xz",
      "versionInfo": "5.10.186",
      "downloadLocation": "https://www.kernel.org/pub/linux/kernel/v5.x/linux-5.10.186.tar.xz",
      "licenseConcluded": "GPL-2.0-or-later",
      "licenseDeclared": "UNKNOWN AND GPL-2.0-or-later AND GPL-2.0",
      "copyrightText": "",
      "externalRefs": [
        {
          "referenceCategory": "PACKAGE-MANAGER",
          "referenceType": "purl",
          "referenceLocator": "pkg:generic/generic?download_url=https://www.kernel.org/pub/linux/kernel/v5.x/linux-5.10.186.tar.xz"
        }
      ]
    },
```

We can keep doing that, if we track the actual source. But at this point, that isn't actually the source. Our build process relies on a separate source (the `eve-kernel` repo). So my thought is that to reference the upstream source would be incorrect, as that is not where we are getting the source for this. It is coming from that repo. So I think the reference locator and download location should be that repo. Obviously, we include the commit, etc.

Do you agree?

**Comment by eriknordmark** [comment]
> We can keep doing that, if we track the actual source. But at this point, that isn't actually the source. Our build process relies on a separate source (the `eve-kernel` repo). So my thought is that to reference the upstream source would be incorrect, as that is not where we are getting the source for this. It is coming from that repo. So I think the reference locator and download location should be that repo. Obviously, we include the commit, etc.

The source is the eve-kernel repo itself, which is a fork of the upstream. I don't know if it makes sense to extract which version/hash it is forked from, and then reference eve-kernel itself for any patches which we add.
Also, @mikem-zed @rene how likely are we to use different upstream versions for different branches in eve-kernel?

**Comment by mikem-zed** [comment]
> > We can keep doing that, if we track the actual source. But at this point, that isn't actually the source. Our build process relies on a separate source (the `eve-kernel` repo). So my thought is that to reference the upstream source would be incorrect, as that is not where we are getting the source for this. It is coming from that repo. So I think the reference locator and download location should be that repo. Obviously, we include the commit, etc.
> 
> The source is the eve-kernel repo itself, which is a fork of the upstream. I don't know if it makes sense to extract which version/hash it is forked from, and then reference eve-kernel itself for any patches which we add. Also, @mikem-zed @rene how likely are we to use different upstream versions for different branches in eve-kernel?

@eriknordmark this is true, we use different versions. At least nvidia kernel uses different tag a as base
@rouming I dropped a support for  https://github.com/brektrou/rtl8821CU . It is very outdated. If we ever need that we can back port (at least try) from 6.5.x. Finally the support for this device was added.

**Comment by mikem-zed** [comment]
Potentially we can download tarball with a source from github like following `curl -L  https://github.com/lf-edge/eve-kernel/tarball/eve-kernel-amd64-v6.1.38-generic -o eve-kernel-amd64-v6.1.38-generic.tgz`

**Comment by deitch** [comment]
> @deitch I dropped a support for https://github.com/brektrou/rtl8821CU . It is very outdated.

That makes it easier. So all we have is the source in this repo, and zfs. We capture zfs via the dockerfile-add-scanner, so we just need the reference to this repo.

> The source is the eve-kernel repo itself, which is a fork of the upstream

Yes, so we should reference this repo. If we reference a specific commit, it will include all changes, patches, etc.

File: kernel-version.mk
Comment (rene) [inline kernel-version.mk:4]: I would suggesting the following (with fixed indentation and some minor fixes):
```
# This file is included by the main Makefile.
# Two variables are used as input for the kernel build:
#   ZARCH - architecture (amd64, arm64)
#   PLATFORM - platform (generic, nvidia, rt, imx*) based on these two
#              variables we set kernel version, commit hash and branch
#              when new kernel is built and pushed to the corresponding
#              dockerhub
#   KERNEL_COMMIT_xxx variable must be manually updated
```

File: kernel-version.mk
Comment (rene) [inline kernel-version.mk:19]: "check **if** ZARCH is supported"

File: kernel-version.mk
Comment (rene) [inline kernel-version.mk:24]: "check **if** PLATFORM is supported"

File: Makefile
Comment (rene) [inline Makefile:119]: I'm not sure if it makes sense to put this info message here, because if the user defines its own KERNEL_TAG, so the KERNEL_{VERSION,BRANCH,COMMIT} might not correspond to the custom TAG, right? In that case, it can create some confusion and misinformation...

File: kernel-version.mk
Comment (rene) [inline kernel-version.mk:34]: Commenting here just to remember you to update this hash....

File: kernel-version.mk
Comment (rene) [inline kernel-version.mk:43]: Commenting here just to remember you to update these hashes....

**Review by eriknordmark — APPROVED** [review]
LGTM

**Comment by eriknordmark** [comment]
I see riscv has issues:
Run make V=1 PRUNE=1 ZARCH=riscv64 pkgs
Makefile:110: "WARNING: We are assembling an riscv6[4](https://github.com/lf-edge/eve/actions/runs/6539128511/job/17756534993?pr=3483#step:8:5) image on amd64. Things may break."
/home/runner/work/eve/eve/kernel-version.mk:21: *** "Unsupported architecture riscv64".  Stop.
Error: Process completed with exit code 2.

**Comment by mikem-zed** [comment]
> I see riscv has issues: Run make V=1 PRUNE=1 ZARCH=riscv64 pkgs Makefile:110: "WARNING: We are assembling an riscv6[4](https://github.com/lf-edge/eve/actions/runs/6539128511/job/17756534993?pr=3483#step:8:5) image on amd64. Things may break." /home/runner/work/eve/eve/kernel-version.mk:21: *** "Unsupported architecture riscv64". Stop. Error: Process completed with exit code 2.

I do not build riscv kernels for now. I'll add them in the follow up PR

File: kernel-version.mk
Comment (mikem-zed) [reply kernel-version.mk]: fixed and pushed

File: kernel-version.mk
Comment (mikem-zed) [reply kernel-version.mk:24]: fixed

File: kernel-version.mk
Comment (mikem-zed) [reply kernel-version.mk:19]: fixed

File: kernel-version.mk
Comment (mikem-zed) [reply kernel-version.mk:4]: ack

File: Makefile
Comment (mikem-zed) [reply Makefile:119]: ah, yes. I forgot to remove it

**Comment by eriknordmark** [comment]
> I do not build riscv kernels for now. I'll add them in the follow up PR

I don't think we can put it in master if the riscv build fails, since I don't think it will proceed and publish the other images to docker and GH releases (but I could be wrong - we don't publish the riscv ones but I think the build failure fails subsequent steps.)
Thoughts?

**Comment by mikem-zed** [comment]
> > I do not build riscv kernels for now. I'll add them in the follow up PR
> 
> I don't think we can put it in master if the riscv build fails, since I don't think it will proceed and publish the other images to docker and GH releases (but I could be wrong - we don't publish the riscv ones but I think the build failure fails subsequent steps.) Thoughts?

I added a branch for riscv64 kernel and made changes in kernel-versions.mk. I still need to clean it up and modify config a bit. I have problems building eve for riscv64. `make V=1 PRUNE=1 ZARCH=riscv64 pkgs` works fine but I cannot build neither live nor rootfs images. I'm getting following error: 
```
Image docker.io/linuxkit/init:144c9cee8aed9e30a16940f2bf1d3813883aceda arch riscv64 not found in local cache, pulling                  
Error: Failed to build init tarball from docker.io/linuxkit/init:144c9cee8aed9e30a16940f2bf1d3813883aceda: Could not pull image docker
```
am I missing something?

**Comment by mikem-zed** [comment]
> > > I do not build riscv kernels for now. I'll add them in the follow up PR
> > 
> > 
> > I don't think we can put it in master if the riscv build fails, since I don't think it will proceed and publish the other images to docker and GH releases (but I could be wrong - we don't publish the riscv ones but I think the build failure fails subsequent steps.) Thoughts?
> 
> I added a branch for riscv64 kernel and made changes in kernel-versions.mk. I still need to clean it up and modify config a bit. I have problems building eve for riscv64. `make V=1 PRUNE=1 ZARCH=riscv64 pkgs` works fine but I cannot build neither live nor rootfs images. I'm getting following error:
> 
> ```
> Image docker.io/linuxkit/init:144c9cee8aed9e30a16940f2bf1d3813883aceda arch riscv64 not found in local cache, pulling                  
> Error: Failed to build init tarball from docker.io/linuxkit/init:144c9cee8aed9e30a16940f2bf1d3813883aceda: Could not pull image docker
> ```
> 
> am I missing something?

ok, **HV=mini** was missing. I'll work on riscv branch for kernel

**Comment by rouming** [comment]
@mikem-zed @eriknordmark do we really need builds for riscv arch? What if we kill it from a pipeline in order to speed up the process so get this PR merged and won't invest time into the riscv until there is a demand

**Comment by eriknordmark** [comment]
> @mikem-zed @eriknordmark do we really need builds for riscv arch? What if we kill it from a pipeline in order to speed up the process so get this PR merged and won't invest time into the riscv until there is a demand

You could remove it from eve-kernel, but then you also need to remove it from various GH workflows in eve as well. Plus remember to add it back once we can build riscv kernels.

**Comment by shjala** [comment]
@mikem-zed 
Sorry I'm late to the party, is there any plan to enabled the module signing and reproducible builds (the go hand in hand)?

**Pull request merged**
